File: Kalorie/Kalorie/Core/UseCases/FetchMealTypesUseCase.swift
//
//  FetchMealTypesUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

protocol FetchMealTypesUseCaseProtocol {
    func callAsFunction() async throws -> [MealTypeDomain]
}

struct FetchMealTypesUseCase: FetchMealTypesUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [MealTypeDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dtos: [MealTypeDTO] = try await dataProvider.loadAsync(from: Constants.Firestore.mealTypes(userId: userId))
        let calendar = Calendar.current
        let today = Date.now
        return dtos
            .compactMap { dto -> MealTypeDomain? in
                guard
                    let start = calendar.date(bySettingHour: dto.startMinutes / 60, minute: dto.startMinutes % 60, second: 0, of: today),
                    let end = calendar.date(bySettingHour: dto.endMinutes / 60, minute: dto.endMinutes % 60, second: 0, of: today)
                else { return nil }
                return MealTypeDomain(id: dto.id, name: dto.name, startTime: start, endTime: end)
            }
            .sorted { $0.startTime < $1.startTime }
    }
}

#if DEBUG
struct FetchMealTypesUseCaseFake: FetchMealTypesUseCaseProtocol {

    // MARK: - Functions

    func callAsFunction() async throws -> [MealTypeDomain] { [] }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchMealTypesUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MealTypeDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants
import java.time.Duration
import java.time.Instant
import java.time.ZoneId

interface FetchMealTypesUseCaseProtocol {
    suspend operator fun invoke(): List<MealTypeDomain>
}

class FetchMealTypesUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchMealTypesUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<MealTypeDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dtos: List<MealTypeDTO> = dataProvider.loadAsync(Constants.Firestore.mealTypes(userId))
        val zone = ZoneId.systemDefault()
        val dayStart = Instant.now().atZone(zone).toLocalDate().atStartOfDay(zone).toInstant()
        return dtos
            .map { dto ->
                MealTypeDomain(
                    id = dto.id,
                    name = dto.name,
                    startTime = dayStart.plus(Duration.ofMinutes(dto.startMinutes.toLong())),
                    endTime = dayStart.plus(Duration.ofMinutes(dto.endMinutes.toLong())),
                )
            }
            .sortedBy { it.startTime }
    }
}
